src/pycapsule.rs

```diff
@@ -0,0 +1,414 @@
+    pub unsafe fn import<'py, T>(py: Python<'py>, name: &CStr) -> PyResult<&'py T> {
+        let ptr = ffi::PyCapsule_Import(name.as_ptr(), false as c_int);
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(&*(ptr as *const T))
+        }
+    }
+
+    /// Sets the context pointer in the capsule.
+    ///
+    /// # Notes
+    ///
+    /// Context itself, is treated much like the value of the capsule, but should likely act as
+    /// a place to store any state managment when using the capsule.
+    ///
+    /// If you want to store a Rust value as the context, and drop it from the destructor, use
+    /// `Box::into_raw` to convert it into a pointer, see the example.
+    ///
+    /// Finally, if `set_context` is called twice in a row, the previous value is always leaked.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use std::ffi::CString;
+    /// use std::sync::mpsc::{channel, Sender};
+    /// use libc::c_void;
+    /// use pyo3::{prelude::*, pycapsule::PyCapsule};
+    ///
+    /// let (tx, rx) = channel::<String>();
+    ///
+    /// fn destructor(val: u32, context: *mut c_void) {
+    ///     let ctx = unsafe { *Box::from_raw(context as *mut Sender<String>) };
+    ///     ctx.send("Destructor called!".to_string()).unwrap();
+    /// }
+    ///
+    /// Python::with_gil(|py| {
+    ///     let name = CString::new("foo").unwrap();
+    ///     let capsule =
+    ///         PyCapsule::new_with_destructor(py, 123, &name, destructor as fn(u32, *mut c_void))
+    ///             .unwrap();
+    ///     let context = Box::new(tx);  // `Sender<String>` is our context, box it up and ship it!
+    ///     capsule.set_context(py, Box::into_raw(context) as *mut c_void).unwrap();
+    ///     // This scope will end, causing our destructor to be called...
+    /// });
+    ///
+    /// assert_eq!(rx.recv(), Ok("Destructor called!".to_string()));
+    /// ```
+    #[allow(clippy::not_unsafe_ptr_arg_deref)]
+    pub fn set_context(&self, py: Python, context: *mut c_void) -> PyResult<()> {
+        let result = unsafe { ffi::PyCapsule_SetContext(self.as_ptr(), context) as u8 };
+        if result != 0 {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Gets the current context stored in the capsule. If there is no context, the pointer
+    /// will be null.
+    pub fn get_context(&self, py: Python) -> PyResult<*mut c_void> {
+        let ctx = unsafe { ffi::PyCapsule_GetContext(self.as_ptr()) };
+        if ctx.is_null() && self.is_valid() && PyErr::occurred(py) {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(ctx)
+        }
+    }
+
+    /// Obtains a reference to the value of this capsule.
+    ///
+    /// # Safety
+    ///
+    /// It must be known that this capsule's pointer is to an item of type `T`.
+    pub unsafe fn reference<T>(&self) -> &T {
```

Comment:
Should this be called `as_ref` and add a `as_mut` companion if applicable? ([`NonNull`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_ref) does something like this).

Reply:
I'm inclined to agree with https://github.com/PyO3/pyo3/pull/1980#discussion_r753810681 and just not have this function - leave it up to the user to handle the pointers.

Reply:
I'm not heavily bound to it, however, it seems if basically every use of fetching the value is `&*(self.get_pointer() as *const T)`, why not make it slightly easier on the user/reader? If they'd like to treat it differently, `PyCapsule::get_pointer` is available.

Reply:
> Should this be called as_ref and add a as_mut companion if applicable? (NonNull does something like this).

No objection to changing it, if that's the consensus.

Reply:
I vote for keeping the `reference` name and returning a `&T` - `as_ref` usually is a cheap pointer conversion and not unsafe.
`as_mut` is not safe in any case once the capsule is created, since the capsule can be shared like other PyObjects.

In order to be able to also use capsules with arbitrary pointers, `get_pointer` is also needed. Maybe it should be named just `pointer` for consistency.

Reply:
Sounds good to me. Have updated `get_pointer` -> `pointer`